src/typing_extensions.py

```diff
@@ -2646,3 +2647,93 @@ def __or__(self, other):
 
             def __ror__(self, other):
                 return typing.Union[other, self]
+
+
+if hasattr(typing, "TypeAliasType"):
+    TypeAliasType = typing.TypeAliasType
+else:
+    class TypeAliasType:
+        """Create named, parameterized type aliases.
+
+        This provides a backport of the new `type` statement in Python 3.12:
+
+            type ListOrSet[T] = list[T] | set[T]
+
+        is equivalent to:
+
+            T = TypeVar("T")
+            ListOrSet = TypeAliasType("ListOrSet", list[T] | set[T], type_params=(T,))
+
+        The name ListOrSet can then be used as an alias for the type it refers to.
+
+        The type_params argument should contain all the type parameters used
+        in the value of the type alias. If the alias is not generic, this
+        argument is omitted.
+
+        Static type checkers should only support type aliases declared using
+        TypeAliasType that follow these rules:
+
+        - The first argument (the name) must be a string literal.
+        - The TypeAliasType instance must be immediately assigned to a variable
+          of the same name. (For example, 'X = TypeAliasType("Y", int)' is invalid,
+          as is 'X, Y = TypeAliasType("X", int), TypeAliasType("Y", int)').
+
+        """
+
+        def __init__(self, name: str, value, *, type_params=()):
+            if not isinstance(name, str):
+                raise TypeError("TypeAliasType name must be a string")
+            self.__value__ = value
+            self.__type_params__ = type_params
+
+            parameters = []
+            for type_param in type_params:
+                if isinstance(type_param, TypeVarTuple):
+                    parameters.extend(type_param)
+                else:
+                    parameters.append(type_param)
+            self.__parameters__ = tuple(parameters)
+            def_mod = _caller()
+            if def_mod != 'typing_extensions':
+                self.__module__ = def_mod
+            # Setting this attribute closes the TypeAliasType from further modification
+            self.__name__ = name
+
+        def __setattr__(self, __name: str, __value: object) -> None:
+            if hasattr(self, "__name__"):
+                raise TypeError("TypeAliasType cannot be modified")
```

Comment:
Shouldn't it be an `AttributeError`? I feel like having `__setattr__` or `__getattr__` raising anything that's not `AttributeError` or a subclass is kind-of evil. CPython `main` also raises `AttributeError`s:

```pycon
>>> type T = int | str
>>> T.__name__ = 'foo'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: readonly attribute
>>> T.bar = 'aba'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'typing.TypeAliasType' object has no attribute 'bar'
```

```suggestion
                raise AttributeError(f"Can't set attribute {__name!r} on an instance of TypeAliasType")
```

Reply:
We could possibly also add a `__delattr__` method as well, so these attributes can't be deleted

Reply:
Good points, pushed both